Collect RX and TX timing statistics

jukkar opened the pull request:
Calculate how long it takes for RX packet from network driver to application. Same for TX packet from application to network driver. Show this information in net-shell for `net stats` command.
Application needs to enable `CONFIG_NET_PKT_RXTIME_STATS=y` and `CONFIG_NET_PKT_TXTIME_STATS=y` to get this information.

If we go with this route, then I will probably remove `CONFIG_NET_CONTEXT_TIMESTAMP` support as it is conflicting with the `CONFIG_NET_PKT_TXTIME_STATS` (the code looks a bit ugly because of this).

jukkar committed:
net: Add statistics how long packets have spent in RX path

Calculate how long on average net_pkt has spent on its way from
network device driver to the application. The data is only
calculated for UDP and TCP network packets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print RX packet timings

Print information about how long network packets took from
network device driver to the application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Add statistics how long packets have spent in TX path

Calculate how long on average net_pkt has spent on its way from
application to the network device driver. The data is calculated
for all network packets and not just for UDP or TCP ones like in
RX statistics.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print TX packet timings

Print information about how long network packets took from
application to the network device driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1187,7 +1187,26 @@ static struct net_pkt *pkt_alloc(struct k_mem_slab *slab, s32_t timeout)
+	if (&tx_pkts == slab) {
+		net_pkt_set_priority(pkt, CONFIG_NET_TX_DEFAULT_PRIORITY);
+	} else if (&rx_pkts == slab) {
+#if defined(CONFIG_NET_PKT_RXTIME_STATS)
```
use IS_ENABLED()

tbursztyka commented on subsys/net/ip/net_stats.h:
```diff
@@ -340,6 +340,28 @@ static inline void net_stats_update_tx_time(struct net_if *iface,
+	u32_t diff = abs(end_time - start_time);
+
+	UPDATE_STAT(iface, stats.rx_time.time_sum +=
+		    SYS_CLOCK_HW_CYCLES_TO_NS64(diff) / 1000);
```
s/1000/MSEC_PER_SEC or?

jukkar replied:
here we want to get us from ns

tbursztyka replied:
NSEC_PER_USEC  yes

tbursztyka commented on subsys/net/ip/net_stats.h:
```diff
@@ -387,6 +409,34 @@ static inline void net_stats_update_tc_tx_time(struct net_if *iface,
+	u32_t diff = abs(end_time - start_time);
+
+	UPDATE_STAT(iface, stats.tc.recv[tc].rx_time.time_sum +=
+		    SYS_CLOCK_HW_CYCLES_TO_NS64(diff) / 1000);
```
same here

tbursztyka commented on include/net/net_stats.h:
```diff
@@ -225,6 +233,7 @@ struct net_stats_tc {
 		net_stats_t pkts;
 		net_stats_t bytes;
 		u8_t priority;
+		struct net_stats_rx_time rx_time;
```
you could probably reorder, to put priority at the end. (looks like it's same in sent array.

tbursztyka commented on subsys/net/ip/net_stats.h:
```diff
@@ -387,6 +409,34 @@ static inline void net_stats_update_tc_tx_time(struct net_if *iface,
+	net_stats_update_rx_time(iface, start_time, end_time);
+}
+#else
+static inline void net_stats_update_tc_rx_time(struct net_if *iface,
```
now that I look at the signature, maybe it would be much simpler by providing only one parameter: net_pkt.
From here you can get the iface, its priority, the start timestamp, the end can be directly got from k_cycle_get(). 

and that will help for (see below):

jukkar replied:
The RX API is done similar way as the TX one and I would like to keep them like that. For the TX one, the net_pkt is freed before the stats function is called, and I would like not to have net_pkt_ref() called as it keeps net_pkt in use for longer and is generally pain to debug.
But I think we can overcome this by not using inline function for the update function, if we have just a define, then the parameters will just disappear and we can avoid the ifdef's. I will investigate this route more.

tbursztyka replied:
indeed, I checked tx side and that would need a bit of rework to get to a simpler signature. As you want then.
That said, why are we counting tx stats only if there is a context attached to it? (does that mean we calculate the stats only on user packets?)

jukkar replied:
The context time counting was done earlier, the idea was that application could activate statistics counting for just one socket. With this new config option, we can get the statistics for all the sockets. Initially I thought that we could remove the context time counting but after some more thinking, it might make sense to allow the application to gather TX statistics for just one socket so we could keep this functionality. The code looks a bit ugly in places (mainly in net_pkt.h) because of various ifdef's, but we can probably live with that.

tbursztyka commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -765,6 +766,13 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 		return -1;
 	}
 
+#if defined(CONFIG_NET_PKT_RXTIME_STATS)
```
... that will help to avoid these ifdef. (and the line will be much shorter as well).

pfalcon commented on include/net/net_stats.h:
```diff
@@ -222,6 +230,7 @@ struct net_stats_tc {
 	} sent[NET_TC_TX_COUNT];
 
 	struct {
+		struct net_stats_rx_time rx_time;
```
But this means that this is cumulative statistics, right? Nowhere does commit message makes this clear. The way it's worded, it seems that you collect per-packet stats.

pfalcon commented on include/net/net_stats.h:
```diff
@@ -210,6 +210,14 @@ struct net_stats_tx_time {
+ */
+struct net_stats_rx_time {
+	u64_t time_sum;
+	net_stats_t time_count;
```
I wouldn't guess what "time_count" mean. Can this be just "count"?

pfalcon commented on subsys/net/ip/Kconfig:
```diff
@@ -640,6 +648,18 @@ config NET_PKT_TXTIME
+	select NET_STATISTICS
+	depends on (NET_UDP || NET_TCP)
+	help
+	  Enable network packet RX time statistics support. This can be used to
```
Also no clear mentioning that the statistics is cumulative for all packets received.

pfalcon commented on subsys/net/ip/net_stats.h:
```diff
@@ -340,6 +340,21 @@ static inline void net_stats_update_tx_time(struct net_if *iface,
+					    u32_t start_time,
+					    u32_t end_time)
+{
+	u32_t diff = abs(end_time - start_time);
```
This is incorrect way to calculate time difference. (Standard comment to your code like that, eh?)

jukkar replied:
> (Standard comment to your code like that, eh?)

That was also a standard comment from you, eh?

pfalcon replied:
No, I multiple times suggested, that when dealing with times represented as unsigned integers, one of them known to be physically after another, then `after - before` is the correct way to calculate the duration, including supplying a Linux test sample which shows correctness for boundary cases, etc. So, it's me who got used to your ignoring the review comments and shoving problematic code into the tree, due to which attitude the network stack stays broken for years.

jukkar replied:
Instead of whining about this, just propose a one liner that does what you expect.
The code currently in the PR works just fine, just tested it in Linux.

pfalcon replied:
Please be ready to accept criticism in you PRs without calling it whining. (And yeah, my point is that this issue was raised enough times to be remembered, but maybe it's just me.)

The correct code is `u32_t diff = end_time - start_time;`.

> The code currently in the PR works just fine, just tested it in Linux.

Well, so what your Linux calculates for the case of start_time = 2, end_time = 1 ?

jukkar replied:
Yep, good point. You could have proposed a proper fix in your first comment so we could avoid this ping pong.

pfalcon replied:
> You could have proposed a proper fix in your first comment so we could avoid this ping pong.

Noted, thanks. Will try to look at this stuff as if it is new and doesn't have (up to) 3 year-history behind it.

pfalcon commented on subsys/net/ip/net_shell.c:
```diff
@@ -783,6 +783,31 @@ static void print_tc_rx_stats(const struct shell *shell, struct net_if *iface)
 	PR("RX traffic class statistics:\n");
+
+#if defined(CONFIG_NET_PKT_RXTIME_STATS)
+	PR("TC  Priority\tRecv pkts\tbytes\ttime\n");
```
Would it make sense to specify time unit straight in the header here?

pfalcon replied:
On closer look, I see that each table row has "us" for that column (code below). Not sure which way is better, but that addresses this concern.

pfalcon reviewed (commented):
Both commit messages have passage like:

> Calculate how long net_pkt has spent *from* its way from application
to the network device driver.

Sounds confusing, should be:

> Calculate how long net_pkt has spent *on* its way from application
to the network device driver.


(And packets should be in plural, and "cumulative time" should be mentioned explicitly, as suggested before.)

jukkar committed:
net: stats: Add functions to reset the statistics

This function will be useful in shell when we want to monitor
the amount of bytes transferred and want to clear earlier
statistics.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Add command to clear network statistics

Needed when testing the amount of bytes transferred.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
> If we go with this route, then I will probably remove CONFIG_NET_CONTEXT_TIMESTAMP support as it is conflicting with the CONFIG_NET_PKT_TXTIME_STATS (the code looks a bit ugly because of this).

So, is it CONFIG_NET_CONTEXT_TIMESTAMP which risks being removed, or CONFIG_NET_PKT_TXTIME ? (Based on this comment https://github.com/zephyrproject-rtos/zephyr/pull/19690/commits/3e60855d1370a04b8d6e22833ff50eb1f746d9ca#diff-8983724739384d9f38274725cd16e2cfR100 "We do not support combination of TXTIME and TXTIME_STATS as the same variable is shared in net_pkt.h").

As was clarified, this patch provides cumulative (averaged) timing stats, whereas CONFIG_NET_PKT_TXTIME records per-packet timing. Both have their usecase, and I'd say losing CONFIG_NET_PKT_TXTIME would be, well, a loss. OTOH, description of CONFIG_NET_CONTEXT_TIMESTAMP shows quite an adhoc feature ("It is possible to add information when the outgoing network packet should be sent."), no idea how precise that description and how that would be implemented.

pfalcon commented on subsys/net/ip/net_stats.h:
```diff
@@ -318,27 +318,21 @@ static inline void net_stats_update_ipv6_mld_drop(struct net_if *iface)
 					    u32_t start_time,
 					    u32_t end_time)
 {
-	u32_t diff = abs(end_time - start_time);
```
Oh, thanks for fixing for the older cases. You might see why I was sour with my comment - because that issue already surfaced (I definitely remember it), and that time, my suggestions were of no avail.

pfalcon reviewed (approved):
Thanks for updates based on the review comments, looks good to me.

jukkar commented on subsys/net/ip/net_shell.c:
```diff
@@ -3462,7 +3462,11 @@ static int cmd_net_stats(const struct shell *shell, size_t argc, char *argv[])
 
-	cmd_net_stats_iface(shell, argc, argv);
+	if (strcmp(argv[1], "reset") == 0) {
+		net_stats_reset(NULL);
```
Yes, that is for further commits.
